theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -167,31 +204,17 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
             cuda_exit(c->ctx);
             return 1;
           }
+          prev_algo.algo = algo;
+          // no tensor_op returned from Get()
+          prev_algo.mathType = mathtype = CUDNN_DEFAULT_MATH;
+	  // fprintf(stderr, "(cudnnGetConvolutionForwardAlgorithm: (err:%d), algo: %d\n", err, algo);
       }
-      prev_algo = algo;
-    } else {
-      algo = prev_algo;
+    } else { 
+      algo = (cudnnConvolutionFwdAlgo_t)prev_algo.algo;
+      worksize = prev_algo.wsSize;
+      mathtype = prev_algo.mathType;
     }
-
-    #ifdef DEBUG
-    if (0 != theano_enum_to_string_cudnnConvolutionFwdAlgo_t(algo, algorithm_name))
-        return 1;
-    // NB: This is printed only when algorithm is chosen at runtime.
-    if (reuse_algo)
-        fprintf(stderr, "(reused %s)\n", algorithm_name);
-    else
-        fprintf(stderr, "(using %s)\n", algorithm_name);
-    #endif
-
-    if (params->choose_once) {
-      reuse_algo = 1;
-    } else {
-      for (unsigned int i = 0; i < PyGpuArray_NDIM(input); i++) {
-        prev_img_dims[i] = PyGpuArray_DIM(input, i);
-        prev_kern_dims[i] = PyGpuArray_DIM(kerns, i);
-      }
-    }
-  }
+  } else { /* choose_algo */
```

Comment:
I would disagree. Sections like that may create false impression for the user that the algo selected actually works. If there is a CUDNN bug/limitation, Theano should get CUDNN_NOT_SUPPORTED and report it to the user (with additional info how to work around) so that the workaround ends up in user's code. 
Can we put this code in a fallback section, like:
 if (status==CUDNN_NOT_SUPPORTED || ! params->choose_algo) ?